patches/server/0977-Add-mob-spawner-ranges-changes-and-API.patch

```diff
@@ -0,0 +1,108 @@
+ import java.util.function.Function;
+ import javax.annotation.Nullable;
++
++import io.papermc.paper.configuration.GlobalConfiguration;
+ import net.minecraft.core.BlockPos;
+ import net.minecraft.core.particles.ParticleTypes;
+ import net.minecraft.core.registries.BuiltInRegistries;
+@@ -44,7 +46,8 @@ public abstract class BaseSpawner {
+     private Entity displayEntity;
+     public int maxNearbyEntities = 6;
+     public int requiredPlayerRange = 16;
+-    public int spawnRange = 4;
++    public int spawnRange = GlobalConfiguration.get().misc.mobSpawnerRange.horizontal; // Paper
++    public int spawnRangeVertical = GlobalConfiguration.get().misc.mobSpawnerRange.vertical; // Paper
+     private int tickDelay = 0; // Paper
+ 
+     public BaseSpawner() {}
+@@ -110,7 +113,7 @@ public abstract class BaseSpawner {
+                     ListTag nbttaglist = nbttagcompound.getList("Pos", 6);
+                     int j = nbttaglist.size();
+                     double d0 = j >= 1 ? nbttaglist.getDouble(0) : (double) pos.getX() + (randomsource.nextDouble() - randomsource.nextDouble()) * (double) this.spawnRange + 0.5D;
+-                    double d1 = j >= 2 ? nbttaglist.getDouble(1) : (double) (pos.getY() + randomsource.nextInt(3) - 1);
++                    double d1 = j >= 2 ? nbttaglist.getDouble(1) : (double) (pos.getY() + spawnRangeVertical == 0 ? 0 : (spawnRangeVertical < 0 ? randomsource.nextIntBetweenInclusive(spawnRangeVertical, -spawnRangeVertical) : randomsource.nextIntBetweenInclusive(-spawnRangeVertical, spawnRangeVertical))); // Paper
+                     double d2 = j >= 3 ? nbttaglist.getDouble(2) : (double) pos.getZ() + (randomsource.nextDouble() - randomsource.nextDouble()) * (double) this.spawnRange + 0.5D;
+ 
+                     if (world.noCollision(((EntityType) optional.get()).getAABB(d0, d1, d2))) {
+@@ -161,7 +164,7 @@ public abstract class BaseSpawner {
+                             return;
+                         }
+ 
+-                        int k = world.getEntitiesOfClass(entity.getClass(), (new AABB((double) pos.getX(), (double) pos.getY(), (double) pos.getZ(), (double) (pos.getX() + 1), (double) (pos.getY() + 1), (double) (pos.getZ() + 1))).inflate((double) this.spawnRange)).size();
++                        int k = world.getEntitiesOfClass(entity.getClass(), (new AABB((double) pos.getX(), (double) pos.getY(), (double) pos.getZ(), (double) (pos.getX() + 1), (double) (pos.getY() + 1), (double) (pos.getZ() + 1))).inflate((double) this.spawnRange, (double) spawnRangeVertical, (double) spawnRange)).size(); // Paper
+ 
+                         if (k >= this.maxNearbyEntities) {
+                             this.delay(world, pos);
+@@ -290,6 +293,12 @@ public abstract class BaseSpawner {
+         if (nbt.contains("SpawnRange", 99)) {
+             this.spawnRange = nbt.getShort("SpawnRange");
+         }
++    
++        // Paper start
++        if (nbt.contains("SpawnRangeVertical", 99)) {
```

Comment:
Prefix the key with "Paper." so there is no chance if a key conflict in the future. Also it'd be good to make a private static final String field with the key.